src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -2087,28 +2080,45 @@ private String initializeDeviceId() {
 
         // see if device id already stored in db
         String deviceId = dbHelper.getValue(DEVICE_ID_KEY);
+        String sharedPrefDeviceId = Utils.getStringFromSharedPreferences(context, instanceName, DEVICE_ID_KEY);
         if (!(Utils.isEmptyString(deviceId) || invalidIds.contains(deviceId))) {
+            // compare against device id stored in backup storage and update if necessary
+            if (!deviceId.equals(sharedPrefDeviceId)) {
+                Utils.writeStringToSharedPreferences(context, instanceName, DEVICE_ID_KEY, deviceId);
+            }
+
             return deviceId;
         }
 
+        // backup #1: check if device id is stored in shared preferences
+        if (!(Utils.isEmptyString(sharedPrefDeviceId) || invalidIds.contains(sharedPrefDeviceId))) {
+            dbHelper.insertOrReplaceKeyValue(DEVICE_ID_KEY, sharedPrefDeviceId);
+            return sharedPrefDeviceId;
```

Comment:
I think these two cases are correct, but I kind of like the idea of all paths of this method ending up calling `saveDeviceId` just so it's easy to verify that we never put ourselves in an inconsistent state.

Reply:
Good suggestion